File: src/stdune/fcntl.c
#include <fcntl.h>
#include <stdio.h>

#include <caml/custom.h>
#include <caml/mlvalues.h>
#include <caml/fail.h>
#include <caml/alloc.h>
#include <caml/memory.h>

CAMLprim value fcntl_lk_native(value fd, value op_v, value type_v, value whence_v, value start, value len)
{
  int op;
  switch (Int_val(op_v))
  {
  case 0:
    op = F_SETLK;
    break;
  case 1:
    op = F_SETLKW;
    break;
  case 2:
    op = F_GETLK;
    break;
  default:
    caml_failwith("fcntl: invalid lock operation");
  }
  int type;
  switch (Int_val(type_v))
  {
  case 0:
    type = F_RDLCK;
    break;
  case 1:
    type = F_WRLCK;
    break;
  case 2:
    type = F_UNLCK;
    break;
  default:
    caml_failwith("fcntl: invalid lock type");
  }
  int whence;
  switch (Int_val(whence_v))
  {
  case 0:
    whence = SEEK_SET;
    break;
  case 1:
    whence = SEEK_CUR;
    break;
  case 2:
    whence = SEEK_END;
    break;
  default:
    caml_failwith("fcntl: invalid lock operation");
  }
  struct flock l = {
    .l_type = type,
    .l_whence = whence,
    .l_start = Long_val(start),
    .l_len = Long_val(len),
  };

  int rv = fcntl(Int_val(fd), op, &l);

  value res;

  res = caml_alloc(2, 0);

  if (op == F_GETLK)
  {
    switch (l.l_type)
    {
      case F_UNLCK:
        Store_field(res, 0, Val_int(0));
        Store_field(res, 1, Val_int(0));
        break;
      case F_RDLCK:
        Store_field(res, 0, Val_int(1));
        Store_field(res, 1, Val_int(l.l_pid));
        break;
      case F_WRLCK:
        Store_field(res, 0, Val_int(2));
        Store_field(res, 1, Val_int(l.l_pid));
        break;
      default:
        Store_field(res, 0, Val_int(-1));
        Store_field(res, 1, Val_int(0));
        break;
    }
  }
  else
  {
    Store_field(res, 0, Val_int(rv));
    Store_field(res, 1, Val_int(0));
  }
  return res;
}

CAMLprim value fcntl_lk_bytecode(value* argv, int argn)
{
  return fcntl_lk_native(argv[0], argv[1], argv[2], argv[3], argv[4], argv[5]);
}

File: src/stdune/fcntl.ml
type lock_type =
  | Read
  | Write

type lock_operation =
  | F_SETLK
  | F_SETLKW
  | F_GETLK

type l_type =
  | F_RDLCK
  | F_WRLCK
  | F_UNLCK

type whence =
  | Set
  | Cur
  | End

external fcntl_lk :
     Unix.file_descr
  -> lock_operation
  -> l_type
  -> whence
  -> int
  -> int
  -> int * int = "fcntl_lk_bytecode" "fcntl_lk_native"

let of_type = function
  | Read -> F_RDLCK
  | Write -> F_WRLCK

let lock ?(whence = Set) ?(start = 0) ?(len = 0) fd t =
  match fcntl_lk fd F_SETLKW (of_type t) whence start len with
  | -1, _ -> raise (Unix.Unix_error (Unix.EINTR, "fcntl", "F_SETLKW"))
  | 0, _ -> ()
  | r, _ ->
    failwith (Printf.sprintf "unexpected fcntl(F_SETLKW) return code: %i" r)

let lock_try ?(whence = Set) ?(start = 0) ?(len = 0) fd t =
  match fcntl_lk fd F_SETLK (of_type t) whence start len with
  | -1, _ -> false
  | 0, _ -> true
  | r, _ ->
    failwith (Printf.sprintf "unexpected fcntl(F_SETLK) return code: %i" r)

let lock_get ?(whence = Set) ?(start = 0) ?(len = 0) fd t =
  match fcntl_lk fd F_GETLK (of_type t) whence start len with
  | 0, _ -> None
  | 1, pid -> Some (Read, pid)
  | 2, pid -> Some (Write, pid)
  | r, _ ->
    failwith (Printf.sprintf "unexpected fcntl(F_GETLK) return code: %i" r)

let unlock ?(whence = Set) ?(start = 0) ?(len = 0) fd =
  match fcntl_lk fd F_SETLKW F_UNLCK whence start len with
  | 0, _ -> ()
  | r, _ ->
    failwith
      (Printf.sprintf "unexpected fcntl(F_SETLK, F_UNLCK) return code: %i" r)
